### ISSUE TITLE
Phone Number Validation Message Issue in Login Page 

### ISSUE THREAD

**Issue by Santhosh-Sellavel** [Expensify/App#4415]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open new Expensify.
2. Log out if already logged in
3. Login using a phone number, enter the phone number without country code

## Expected Result:
While entering it should have shown the message 
       Please enter a phone number including the country code e.g +447814266907
       
## Actual Result:
Only after press continue an error message is shown.

### While Entering
<img width="474" alt="Screenshot 2021-08-05 at 6 21 15 AM" src="https://user-images.githubusercontent.com/85645967/128274600-f48db763-1fb1-41a5-836e-b4aaee95579d.png">

### After pressing continue
<img width="386" alt="Screenshot 2021-08-05 at 6 26 17 AM" src="https://user-images.githubusercontent.com/85645967/128274606-a355555d-1e4f-478e-bd32-f089dfffb088.png">

## Workaround:
Yes users can add country code after an error shows up

## Proposal 
We already had a similar issue in search we can use the same check for phoneNumber 

In `LoginForm.js `https://github.com/Expensify/App/blob/474a1492b5507390cf65aa177aa4e20b6f06a0a4/src/pages/signin/LoginForm.js#L81-L90

add these lines below,
```
{this.state.login
          && !CONST.REGEX.DIGITS_AND_PLUS.test(this.state.login) && Str.isValidPhone(this.state.login)
              && (
              <Text style={[styles.formError]}> 
                  {this.props.translate('messages.noPhoneNumber')}
              </Text>
              )}
```

Will change styles based on input from the team.
Also, we can move the phone validation logic to `Str.js` in expensify-common,  if needed, and use the same in other places

______________________________________________________________
## Platforms
Where the issue is occuring?

- [x] Web
- [x] iOS
- [x] Android
- [x] Desktop App
- [x]  Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MitchExpensify** [comment]
Hey @Santhosh-Sellavel! Sorry for the delay. Confirmed the message doesn't show when typing a number without the country code. This seems like a reasonable improvement, labelling eng to review your proposal before they label external and have the Upwork job created

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MitchExpensify** [comment]
Oops I got the order wrong. First I should label external which will cause a job to be created in Upwork. Once done it will be labelled exported for engineering to review the proposal. Unassigning @tgolen and labeling external

**Comment by marklouisdeshaun** [comment]
Created a job in Upwork!

Internal: https://www.upwork.com/ab/applicants/1425538123129339904/job-details
External: https://www.upwork.com/jobs/~0158c04a9aa11f0914

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tauseefnasir** [comment]
@marklouisdeshaun I can fix this issue. All I have to do is add a callback on onChange of input and on each key press I will validate user input against a regex formate and will display message as soon as regex match fails.

**Comment by deetergp** [comment]
The proposal from @Santhosh-Sellavel looks good 👍

**Comment by Santhosh-Sellavel** [comment]
Proposal slightly updated, Kindly verify when you have time. @deetergp

I'll wrap up a PR by tomorrow!

Just one concern regarding style can we leave the same style as below or can we update?

![Screenshot_20210813-133422_Samsung Internet.jpg](https://user-images.githubusercontent.com/85645967/129325136-0b4f5ac1-417d-456e-9796-3705794d807a.jpg)

**Comment by Santhosh-Sellavel** [comment]
@Beamanator
Do I need to hold this, since your are writing requirements for phone number validation regex & this issue is similar too.

Earlier I suggested validation changes from PR https://github.com/Expensify/App/pull/4354, since changes were approved.
But when I validated the solution today it does not work, so only asked here to clarfiy in https://github.com/Expensify/App/issues/2934#issuecomment-898233586
and got clarified about requirement change there.

In here I'm planning to `update existings regex` or` add a new regex` to validate the phone number, show messages.
DIGITS_AND_PLUS Regex will updated or will replace with a new regex  ` /^\+(?:[0-9]*)$/`
```
{this.state.login
          && !CONST.REGEX.DIGITS_AND_PLUS.test(this.state.login) && Str.isValidPhone(this.state.login)
              && (
              <Text style={[styles.formError]}> 
                  {this.props.translate('messages.noPhoneNumber')}
              </Text>
              )}
```
What do you say, do I need to hold this, can I proceed with PR?

**Comment by Beamanator** [comment]
I'd say go for this one, whatever y'all decided here! I was actually talking about a different issue that isn't related to the login page, so I'd say proceed on this one :)

**Comment by Santhosh-Sellavel** [comment]
> I'd say go for this one, whatever y'all decided here! I was actually talking about a different issue that isn't related to the login page, so I'd say proceed on this one :)

Thanks again for clarification! @Beamanator

**Comment by Santhosh-Sellavel** [comment]
cc: @marklouisdeshaun  @deetergp 

After approaching the solution for this issue. 
I found the following issues/concerns.

1. Enter 16 or more numbers, even though this looks like a phone number, we show an "invalid email" message.
<img width="469" alt="Screenshot 2021-08-13 at 7 40 28 PM" src="https://user-images.githubusercontent.com/85645967/129370089-4fa080cb-3446-4edb-b0c6-723a671bb7e0.png">

2. In empty login do we need to allow continue? 
<img width="577" alt="Screenshot 2021-08-13 at 7 45 13 PM" src="https://user-images.githubusercontent.com/85645967/129370703-f4d50156-3532-495c-bdb9-b03fcbd222cd.png">

3. After the user press continue button in empty login we show the error as in above. When the user gives input do we need to persist that error message?
<img width="469" alt="Screenshot 2021-08-13 at 7 50 49 PM" src="https://user-images.githubusercontent.com/85645967/129371836-fb2c27ec-2b7d-4780-8dc3-b42d1ef4844b.png">

It may lead to something like this if we don't clear
<img width="444" alt="Screenshot 2021-08-13 at 7 38 03 PM" src="https://user-images.githubusercontent.com/85645967/129370567-c5244c7d-1e12-4929-98b6-0042573ad302.png">

4. In this issue context, can we allow the user to press the continue button when we already show the number is invalid?
<img width="497" alt="Screenshot 2021-08-13 at 7 38 32 PM" src="https://user-images.githubusercontent.com/85645967/129373193-2396f995-45de-4a6b-9648-ece103728b23.png">

5. Can we show the user multiple errors like this.
<img width="465" alt="Screenshot 2021-08-13 at 8 07 40 PM" src="https://user-images.githubusercontent.com/85645967/129373833-48210735-5e13-441f-8d82-16d3eb2f68b7.png">

Can I separate few issues and raise them separately except(4)?
If so which issues to raise separately or can be group. 

Or We can handle all these issues here?

Post your inputs thanks!

**Comment by deetergp** [comment]
@Santhosh-Sellavel You've definitely brought up some good points in your investigation. Here's my take on it: fix the phone number problem in this issue and include a fix for 4 since it does seem related.

Then create a new issue regarding how the error messages and continue button interact with one another. I think that is a bigger discussion that should stand on its own and require more input from others. Sound good?

**Comment by parasharrajat** [comment]
To me, it seems that all are the result of incorrectly parsing the input value and calculating the error early. cc @Santhosh-Sellavel.

**Comment by Santhosh-Sellavel** [comment]
Only `validation or input parsing` is done here on the client-side is whether login is empty.
In this issue, I will handle phone number validation.
Every other error is calculated from the server-side.

What do you call calculating early?

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel You've definitely brought up some good points in your investigation. Here's my take on it: fix the phone number problem in this issue and include a fix for 4 since it does seem related.
> 
> Then create a new issue regarding how the error messages and continue button interact with one another. I think that is a bigger discussion that should stand on its own and require more input from others. Sound good?

Sure @deetergp!

**Comment by Santhosh-Sellavel** [comment]
@deetergp draft once you are okay with changes, I'll add screenshots and wrap up PR mark it ready for review!

**Comment by marklouisdeshaun** [comment]
Not overdue, just waiting for review of the draft PR

**Comment by deetergp** [comment]
I looked over the draft earlier today and it looked good thus far.

**Comment by deetergp** [comment]
After talking it over with others, we've decided that the way the validation works currently, validating after pressing Continue, is the desired functionality. This is no longer a priority.
